pumpkin-solver/src/propagators/gcc/gcc_extended_resolution/inequality_sets.rs

```diff
@@ -0,0 +1,640 @@
+use rand::seq::SliceRandom;
+use rand::thread_rng;
+
+use crate::basic_types::HashMap;
+use crate::basic_types::HashSet;
+use crate::basic_types::Inconsistency;
+use crate::engine::propagation::LocalId;
+use crate::engine::propagation::PropagationContextMut;
+use crate::engine::propagation::Propagator;
+use crate::engine::propagation::ReadDomains;
+use crate::engine::DomainEvents;
+use crate::predicate;
+use crate::predicates::Predicate;
+use crate::predicates::PropositionalConjunction;
+use crate::variables::IntegerVariable;
+use crate::variables::Literal;
+
+pub(crate) struct GccInequalitySets<Var> {
+    variables: Box<[Var]>,
+    equalities: HashMap<(usize, usize), Literal>,
+}
+
+impl<Var> GccInequalitySets<Var> {
+    pub(crate) fn new(
+        variables: impl IntoIterator<Item = Var>,
+        equalities: HashMap<(usize, usize), Literal>,
+    ) -> Self {
+        Self {
+            variables: variables.into_iter().collect(),
+            equalities,
+        }
+    }
+
+    fn get_equality(&self, x: usize, y: usize) -> Literal {
+        assert!(x < self.variables.len());
+        assert!(y < self.variables.len());
+
+        *self
+            .equalities
+            .get(&(x, y))
+            .or(self.equalities.get(&(y, x)))
+            .expect("E_{x,y} or E_{y,x} must be defined")
+    }
+
+    fn get_inequality_explanation(&self, vars: &[usize]) -> Vec<Predicate> {
+        let mut reason = Vec::new();
+        for i in 0..vars.len() {
+            for j in (i + 1)..vars.len() {
+                let literal = self.get_equality(vars[i], vars[j]);
+                reason.push(!predicate!(literal == 0));
+            }
+        }
+        reason
+    }
+}
+
+impl<Var: IntegerVariable + 'static> Propagator for GccInequalitySets<Var> {
+    fn name(&self) -> &str {
+        "GCC extended resolution inequality sets"
+    }
+
+    fn initialise_at_root(
+        &mut self,
+        context: &mut crate::engine::propagation::PropagatorInitialisationContext,
+    ) -> Result<(), PropositionalConjunction> {
+        // In theory, we could run this propagator for every change in domains of variables
+        // but this would be rather wasteful, right?
```

Comment:
I think we can run a few benchmarks just to see how much time do we spend running this propagator. I think it is better to do all propagations that we can, i.e. never miss propagations.